src/bootstrap/src/ferrocene/run.rs

```diff
@@ -99,3 +100,139 @@ impl Step for TraceabilityMatrix {
         html_output
     }
 }
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub(crate) struct GenerateCoverageReport;
+
+impl Step for GenerateCoverageReport {
+    type Output = ();
+    const ONLY_HOSTS: bool = true;
+    const DEFAULT: bool = true;
+
+    fn run(self, builder: &Builder<'_>) -> Self::Output {
+        builder.info(&format!("Generating coverage report"));
+
+        let coverage_report_data_dir =
+            env_llvm_profile_data_dir().unwrap_or_else(|| builder.out.join("coverage"));
+
+        let coverage_report_out_dir = builder.out.join("coverage_report");
+
+        let coverage_src_path = builder.out.join("../library/core").canonicalize().unwrap();
+        let _cargo_dir = builder.cargo_dir();
+
+        if builder.config.dry_run() {
+            return;
+        }
+
+        let core_tests_binary_path = get_test_binary_path(builder).unwrap();
+
+        if coverage_report_data_dir.exists() {
+            let mut files = coverage_report_data_dir.read_dir().expect("Failed to read dir");
+
+            let has_profraw_files = files.any(|file| {
+                if let Ok(file) = file {
+                    file.path().extension().map(|ext| ext.to_str().unwrap()) == Some("profraw")
+                } else {
+                    false
+                }
+            });
+
+            if !has_profraw_files {
+                panic!("No profraw files found in build/coverage directory");
+            }
+        }
+
+        if coverage_report_out_dir.exists() {
+            builder.remove_dir(&coverage_report_out_dir);
+            builder.info(&format!("Removed previous report at {:?}", coverage_report_out_dir));
+        }
+        let rustc_path = builder.out.join(builder.config.build.triple).join("stage1/bin/rustc");
+
+        let grcov_installed = Command::new("grcov").arg("--version").output().is_ok();
+
+        if !grcov_installed {
+            panic!("grcov not installed, to install it run:\n cargo install grcov");
+        }
+
+        let mut cmd = builder.tool_cmd(Tool::FerroceneGenerateCoverageReport);
+        cmd.env("COVERAGE_REPORT_DATA_DIR", coverage_report_data_dir);
+        cmd.env("COVERAGE_REPORT_OUT_DIR", coverage_report_out_dir);
+        cmd.env("COVERAGE_REPORT_SRC_PATH", coverage_src_path);
+        cmd.env("COVERAGE_REPORT_BIN_PATH", core_tests_binary_path);
+        cmd.env("RUSTC", rustc_path);
```

Comment:
This environment variable doesn't seem to be used by the tool?

Reply:
`RUSTC` environment variable is required by grcov to locate the right version of the `llvm-cov`. If the variable is not set `grcov` uses the the `rustc` in the host. This version mismatch of `llvm-cov` leads to empty report.  So, we need the `stage1/rustc` to create the report.

`grcov` uses `cargo-binutils` to invoke `llvm-cov`. `cargo-binutils` looks for `RUSTC` env variable to get the path for `llvm-cov`.

https://github.com/rust-embedded/cargo-binutils/blob/5c38490e1abf91af51d0a345bb581e37facd28ff/src/rustc.rs#L8.

Reply:
Right, missed it!